core/protocols/aws-json-protocol/src/main/java/software/amazon/awssdk/protocols/json/internal/marshall/QueryParamMarshaller.java

```diff
@@ -60,6 +61,12 @@ public final class QueryParamMarshaller {
         }
     };
 
+    public static final JsonMarshaller<Void> NULL = (val, context, paramName, sdkField) -> {
+        if (sdkField.containsTrait(RequiredTrait.class)) {
+            throw new IllegalArgumentException(String.format("Parameter '%s' must not be null", paramName));
+        }
+    };
```

Comment:
Are we testing required lists and maps bound to query parameters? I'm not sure if it's the list/map or the null marshaller that is used when it's an auto-construct list.

Reply:
Good catch. I'm currently adding tests for list/map. In these instances, a list/map marshaller validates the collection, and then defers to the appropriate query marshaller for the elements. The `SdkField` is not passed, though, so we can't currently validate elements as is. I also noticed lists can't be null as we have guards that default to an empty collection instead.